modules/tensor_mechanics/src/materials/ComputeBlockRotatedElasticityTensor.C

```diff
@@ -0,0 +1,57 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeBlockRotatedElasticityTensor.h"
+#include "EulerAngleProvider.h"
+#include "RotationTensor.h"
+
+registerMooseObject("TensorMechanicsApp", ComputeBlockRotatedElasticityTensor);
+
+template <>
+InputParameters
+validParams<ComputeBlockRotatedElasticityTensor>()
+{
+  InputParameters params = validParams<ComputeElasticityTensorBase>();
+  params.addClassDescription("Compute an elasticity tensor.");
+  params.addRequiredParam<std::vector<Real>>("C_ijkl", "Stiffness tensor for material");
+  params.addParam<MooseEnum>(
+      "fill_method", RankFourTensor::fillMethodEnum() = "symmetric9", "The fill method");
+  params.addRequiredParam<UserObjectName>("euler_angle_provider",
+                                          "Name of Euler angle provider user object");
+  return params;
+}
+
+ComputeBlockRotatedElasticityTensor::ComputeBlockRotatedElasticityTensor(const InputParameters & parameters)
+  : ComputeElasticityTensorBase(parameters),
+
+    _C_ijkl(getParam<std::vector<Real>>("C_ijkl"),
+            (RankFourTensor::FillMethod)(int)getParam<MooseEnum>("fill_method")),
+    _euler(getUserObject<EulerAngleProvider>("euler_angle_provider"))
+{
+}
+
+void
+ComputeBlockRotatedElasticityTensor::computeQpElasticityTensor()
+{
+  EulerAngles angles;
+  auto grain_id = _current_elem->subdomain_id();
+  if (grain_id == 0)
+    mooseError("ZERO!!!!");
+  grain_id--;
+
+  if (grain_id < _euler.getGrainNum())
+    angles = _euler.getEulerAngles(grain_id);
+  else
+    mooseError(
+        "invalid block id ", grain_id, " (only have ", _euler.getGrainNum(), " Eulerangles)");
+
+  RankFourTensor C_ijkl = _C_ijkl;
+  C_ijkl.rotate(RotationTensor(RealVectorValue(angles)));
```

Comment:
You could pre-rotate in `residualSetup` or redesign to rotate in `subdomainSetup` only.

Reply:
Check out `SetupInterface.h` for more info